[TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70475
* __->__ #70375
* #70374



Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

Commit:
[WIP][TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

Commit:
Update on "[WIP][TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

Commit:
Update on "[WIP][TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

Commit:
Update on "[WIP][TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

priyaramani commented on torch/csrc/jit/mobile/nnc/aot_compiler.cpp:
```diff
@@ -189,6 +194,153 @@ std::pair<std::unique_ptr<Function>, const std::string> aotCompile(
+
+  torch::jit::tensorexpr::annotateInputShapes(graph, example_inputs);
+  torch::jit::OptimizeFrozenGraph(graph, true);
+  torch::jit::PropagateShapesOnGraph(graph);
```
Do you think we can remove the shape propagation passes as TraceGraph() is called from aotCompile() ?

ZolotukhinM replied:
I'd prefer to keep it (but I don't feel strongly about it). My motivation is that 1) it doesn't hurt to run shape-prop if all the shapes are already there, 2) in theory it's possible to invoke `preprocess` from some other place than `aotCompile` - in that case shape-prop can actually do work.

priyaramani commented on binaries/aot_model_compiler.cc:
```diff
@@ -217,27 +112,6 @@ int main(int argc, char** argv) {
-    example_inputs.emplace_back(at::rand(inputShapes[i]).to(at::dtype(inputTypes[i])));
-  }
-
-  torch::jit::RemoveTensorMutation(graph);
```
Can we also remove the corresponding headers imported for this and others?

ZolotukhinM replied:
Good point, thanks!

Commit:
Update on "[WIP][TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

Commit:
Update on "[TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

Commit:
Update on "[TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

Commit:
clang-tidy on "[TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

Commit:
clang-tidy on "[TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

ZolotukhinM commented on test/mobile/nnc/test_nnc_backend.cpp:
```diff
@@ -187,49 +95,6 @@ TEST(NNCBackendTest, AOTCompileThenExecute) {
   EXPECT_TRUE(result.toTensor().equal(reference.toTensor()));
 }
 
-TEST(NNCBackendTest, FakeTensor) {
```
I had to remove this test as we won't be able to compile models with calls to methods of custom classes (at least as of now). The other test still exercises the AOT workflow.

Commit:
Update on "[TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

Commit:
Update on "[TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

priyaramani commented on binaries/aot_model_compiler.cc:
```diff
@@ -61,125 +54,20 @@ std::vector<std::string> split(
+  method_spec.insert("sizes", FLAGS_input_dims);
+  method_spec.insert("types", FLAGS_input_types);
+  method_spec.insert("asmfile", FLAGS_output_llvm);
+  method_spec.insert("model_name", FLAGS_model_name);
```
model_name and model_version can be in the compile_spec outside the method_spec ?

ZolotukhinM replied:
There is a bunch of places where we iterate over compile spec and expect keys to be method names, so we can't just put model_name and model_version there. Would you mind if I clean that up in a separate PR after this lands?

priyaramani replied:
Ah that's right there are places where we expect keys to be method names. I'll look at what other delegates are doing and converge them.

priyaramani commented on torch/csrc/jit/mobile/nnc/aot_compiler.cpp:
```diff
@@ -178,40 +183,184 @@ std::pair<std::unique_ptr<Function>, const std::string> aotCompile(
-  GRAPH_DUMP("graph before compiler passes ", g);
-  tensorexpr::removeUnusedSelfArgument(g);
-  g = TraceGraph(g, example_values);
   // TODO: Remove annotateInputShapes pass when TraceGraph can also capture
```
ultra nit: Have the comment with annotateInputShapes removing the empty line in between

Commit:
Update on "[TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

Commit:
Update on "[TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

Commit:
Update on "[TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend"




Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).
